domain/blockdevice/state/state.go

```diff
@@ -288,42 +272,64 @@ func blockDevicesChanged(oldDevices, newDevices []blockdevice.BlockDevice) bool
 	return false
 }
 
-// RemoveMachineBlockDevices removes all the block devices for the specified machine.
-// It is the same as calling SetMachineBlockDevices with an empty list, but does not
-// error if the machine life is Dead.
-func (st *State) RemoveMachineBlockDevices(ctx context.Context, machineId string) error {
+// MachineBlockDevices retrieves block devices for all machines.
+func (st *State) MachineBlockDevices(ctx context.Context) ([]blockdevice.MachineBlockDevice, error) {
 	db, err := st.DB()
 	if err != nil {
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
 
-	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		q := `
-SELECT machine.uuid AS &M.machine_uuid
-FROM   machine
-WHERE  machine.machine_id = $M.machine_id
+	query := `
+SELECT bd.* AS &BlockDevice.*,
+       bdl.* AS &DeviceLink.*,
+       fs_type.* AS &FilesystemType.*,
+       machine.machine_id AS &BlockDeviceMachine.*
+FROM   block_device bd
+       JOIN machine ON bd.machine_uuid = machine.uuid
+       LEFT JOIN block_device_link_device bdl ON bd.uuid = bdl.block_device_uuid
+       LEFT JOIN filesystem_type fs_type ON bd.filesystem_type_id = fs_type.id
 `
-		stmt, err := sqlair.Prepare(q, sqlair.M{})
-		if err != nil {
-			return errors.Trace(err)
-		}
 
-		result := sqlair.M{}
-		err = tx.Query(ctx, stmt, sqlair.M{"machine_id": machineId}).Get(result)
-		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
-			return errors.Annotatef(err, "looking up UUID for machine %q", machineId)
-		}
-		if len(result) == 0 {
-			return fmt.Errorf("machine %q not found%w", machineId, errors.Hide(machineerrors.NotFound))
-		}
-		machineUUID := result["machine_uuid"].(string)
-		if err := RemoveMachineBlockDevices(ctx, tx, machineUUID); err != nil {
-			return errors.Annotatef(err, "removing block devices on machine %q (%s)", machineId, machineUUID)
+	types := []any{
+		BlockDevice{},
+		FilesystemType{},
+		DeviceLink{},
+		BlockDeviceMachine{},
+	}
```

Comment:
This is not nice at all. I wonder if it's more palatable to be outside of the method and just be a package variable.

Reply:
I don't see the issue. The sqlair Prepare() method takes a variadac list of types. It's convenient to assign them to to a var next to the call to Prepare() to keep the code a bit more organised. We do this is a few places.